### ISSUE TITLE
[wayland] Enable NumLock on login

### ISSUE THREAD

**Issue by arjan-s** [qtile/qtile#4225]
### The issue:

Question, how can I enable NumLock by default when using the Wayland backend? It doesn't seem to be implemented in Qtile.

On X, people usually use an external tool like `numlockx`, but this doesn't work on wayland.

For example, Sway does it through their input settings:
https://wiki.archlinux.org/title/Sway#Initially_enable_CapsLock/NumLock
https://github.com/swaywm/sway/blob/master/sway/commands/input/xkb_numlock.c

### Required:

- [X] I have searched past issues to see if this bug has already been reported.

**Comment by Sherry65-code** [comment]
To enable NumLock on login in qtile on Wayland, you can use the following steps:

- Install the numlockx package by running the following command in your terminal:

  ```bash
  sudo apt install numlockx
  ```

- Add the following line to your qtile configuration file (~/.config/qtile/config.py):

  ```python
  import subprocess
  subprocess.call(["numlockx", "on"])
  ```

  This will turn on the NumLock key when qtile starts up.

- Save the configuration file and restart qtile.

  ```bash
  qtile cmd-obj -o cmd -f restart
  ```

After following these steps, the NumLock key should be enabled automatically when you log in to qtile on Wayland.

**Comment by arjan-s** [comment]
Thanks for taking the time to reply to my question, but sadly it isn't very useful. As I said in my original question, numlockx works on X, not on Wayland.

**Comment by Sherry65-code** [comment]
Enabling **NumLock** by default when using the Wayland backend can be tricky, but there are a few things you can try. 

First, you can check if your Wayland compositor (the program that manages your windows and displays) has a setting to enable NumLock by default. If you're using the Sway compositor, you can add a line to your configuration file to enable NumLock for your keyboard.

If your compositor doesn't have this setting, you can create a custom script that runs when you log in to enable NumLock. This involves creating a file with some commands in it that will run automatically when you log in. 

Finally, if neither of those options work, you can try switching to a different keyboard layout that has NumLock enabled by default. The "us" layout is one example. You can switch to this layout with a command in the terminal.

Keep in mind that these solutions may not work for everyone, so you may need to experiment a bit to find a solution that works for you. 😅

**Comment by elParaguayo** [comment]
Qtile is the Wayland compositor.

Pinging @m-col as he might have some ideas about how this is possible and whether the Sway technique (as shown here: https://wiki.archlinux.org/title/Sway#Initially_enable_CapsLock/NumLock) could be replicated with Qtile.

**Comment by arjan-s** [comment]
Honestly I think Sherry65-code is just using ChatGPT to write these answers...

**Comment by Sherry65-code** [comment]
No **Bro**, I am wrote it myself. :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob: :sob:

**Comment by Sherry65-code** [comment]
And if I would use AI, wouldn't I just tell you to use the AI by your own.

**Comment by Sydiepus** [comment]
I did some investigation and from what i learned it seems that there's no "right way" to do it.
[sway](https://github.com/swaywm/sway/pull/2330/files) calls `wlr_keyboard_notify_modifiers` which is a function the compositor is not supposed to use.
Idk how to include this function in pywlroots to make it work.

**Comment by Shinyzenith** [comment]
> I did some investigation and from what i learned it seems that there's no "right way" to do it. [sway](https://github.com/swaywm/sway/pull/2330/files) calls `wlr_keyboard_notify_modifiers` which is a function the compositor is not supposed to use. Idk how to include this function in pywlroots to make it work.

Why is the compositor not supposed to use `wlr_keyboard_notify_modifiers`? Isn't wlroots exported functions literally meant for compositors to begin with?

**Comment by Sydiepus** [comment]
> Why is the compositor not supposed to use `wlr_keyboard_notify_modifiers` ?

This was an assumption on my part mainly backed by the sway PR (the author said it), since this function is not available in the headers files.

> Isn't wlroots exported functions literally meant for compositors to begin with?

Yes they are, however some functions could be internal/helper functions that wlroots uses, which they choose not to make "public". ( as in include them in the header file )

To my understanding, header files are like documentation to which functions/structs you want to be publicly accessible, however there's nothing stopping you from using functions not inside the header file given you know the name and parameters of the function.
I've opened an issue in [pywlroots](https://github.com/flacjacket/pywlroots/issues/121) asking for help regarding this function, and it might not be necessary to implement it since we only need to alter the keyboard state, maybe simulate key press using qtile ? or maybe there's already a way to do that using pywlroots ? maybe even pywayland.

**Comment by Shinyzenith** [comment]
> > Why is the compositor not supposed to use `wlr_keyboard_notify_modifiers` ?
> 
> This was an assumption on my part mainly backed by the sway PR (the author said it), since this function is not available in the headers files.

If I'm not mistaken, sway is just a consumer client of the wlroots API. Given that, how would it use the function if it was not exported by the header files? 

The pr over on the sway side includes `#include <wlr/interfaces/wlr_keyboard.h>` which does have the function defined. If it were to be not exposed by wlr during the time of the merger, we would see a function declaration for `wlr_keyboard_notify_modifiers` in the pr. Correct me if I'm mistaken.

**Comment by Shinyzenith** [comment]
I personally do not see how it is not meant to be used by the compositor. If that were the case, the sway devs would probably be the one to close the PR first? Their implementations of the wayland protocols are pretty much the defacto standards for wlr.

**Comment by Sydiepus** [comment]
Well this embarrassing, while i was working on it i was looking at `types/wlr_keyboard.h` and not `interfaces`.
Really sorry about the inconvenience i caused i should've double checked or even checked the includes.

Now that the function is in a header file i don't see why the compositor is not meant to use the function.
Anyway since sway does it this way we can do it the same way.

Again, sorry for not confirming my information before commenting and thank you for correcting me.

**Comment by Shinyzenith** [comment]
I think it would be wise to just wait for m-col's branch to be merged ( 0.16 ) and rebase your patch against it. No point updating the wlroots 0l15 branch since 0.16 pr is almost ready if I'm not mistaken.

Please do take my advice with a grain of salt, I'm not a member of the qtile team. Maybe @elParaguayo can give a better answer?

**Comment by elParaguayo** [comment]
Agreed. I don't use Wayland but there is a very large PR that's almost done so, really, any other changes should be made after that.

**Comment by Sydiepus** [comment]
Hello, I'm currently working on a fix for this issue, it depends on https://github.com/flacjacket/pywlroots/pull/123

For Qtile would `wl_numlock` and `wl_capslock` be acceptable names for the settings ?
Should I open a pull request or wait for the one in pywlroots to get merged ?

**Comment by otaj** [comment]
Just my two cents here, sway has these options as `xkb_numlock` and `xkb_capslock`

**Comment by Sydiepus** [comment]
Sorry for the extra reference not sure how I messed up branch creation.

**Comment by xz-dev** [comment]
`River` also have the same function missing, so I finding here.
But thanks @Sydiepus , I had create a tool like `numlockx`, hope anyone can test it.
https://github.com/xz-dev/numlockw
